src/storage-client/src/controller/collection_mgmt.rs

```diff
@@ -288,46 +129,211 @@ impl CollectionManager {
+            let appender = self.monotonic_appender(id).expect("id to exist");
+            match appender.append(updates).await {
+                Ok(()) => (),
+                // There's nothing we can do if we're shutting down.
+                Err(StorageError::ShuttingDown(_)) => (),
+                Err(e) => panic!("unexpected error when appending to collection {id}, err: {e:?}"),
+            }
         }
     }
 
     /// Returns a [`MonotonicAppender`] that can be used to monotonically append updates to the
     /// collection correlated with `id`.
-    pub(super) fn monotonic_appender(&self, id: GlobalId) -> MonotonicAppender {
-        MonotonicAppender {
-            id,
-            tx: self.tx.clone(),
-        }
+    pub(super) fn monotonic_appender(
+        &self,
+        id: GlobalId,
+    ) -> Result<MonotonicAppender, StorageError> {
+        let guard = self.collections.lock().expect("CollectionManager panicked");
+        let tx = guard
+            .get(&id)
+            .map(|(tx, _)| tx.clone())
+            .ok_or(StorageError::IdentifierMissing(id))?;
+
+        Ok(MonotonicAppender { tx })
     }
 }
 
+/// Spawns an [`mz_ore::task`] that will continuously bump the upper for the specified collection,
+/// and append data that is sent via the provided [`mpsc::Sender`].
+///
+/// TODO(parkmycar): One day if we want to customize the tick interval for each collection, that
+/// should be done here.
+/// TODO(parkmycar): Maybe add prometheus metrics for each collection?
+fn write_task<T>(
+    id: GlobalId,
+    write_handle: persist_handles::PersistWriteWorker<T>,
+    now: NowFn,
+) -> (WriteChannel, WriteTask)
+where
+    T: Timestamp + Lattice + Codec64 + From<EpochMillis> + TimestampManipulation,
+{
+    let (tx, mut rx) = mpsc::channel(CHANNEL_CAPACITY);
+
+    let handle = mz_ore::task::spawn(
+        || format!("CollectionManager-write_task-{id}"),
+        async move {
+            let mut interval = tokio::time::interval(DEFAULT_TICK);
+
+            'run: loop {
+                tokio::select! {
+                    // Prefer sending actual updates over just bumping the upper, because sending
+                    // updates also bump the upper.
+                    biased;
+
+                    // Pull as many queued updates off the channel as possible.
+                    cmd = rx.recv_many(CHANNEL_CAPACITY) => {
+                        if let Some(batch) = cmd {
+                            // To rate limit appends to persist we add artifical latency, and will
+                            // finish no sooner than this instant.
+                            let min_time_to_complete = Instant::now() + DEFAULT_TICK;
+
+                            // Reset the interval which is used to periodically bump the uppers
+                            // because the uppers will get bumped with the following update.
+                            interval.reset();
```

Comment:
do you mean `min_time_to_complete`? I think reseting the interval and then waiting on it in the select! might be doing the work that the `sleep(min_time_to_complete)` is doing

Reply:
Oh I see the logic is actually more complicated: the sleep is boxing out new updates from arriving. Maybe a comment explaining the expected behavior re times would help (at least help me).

Reply:
Just added a comment to the `interval.reset()` call explaining why it exists, let me know if it makes sense!